data/campaigns/fri05.wmf/scripting/mission_thread.lua

```diff
@@ -0,0 +1,309 @@
+         for _,proposal in ipairs(trade) do
+            local trade_interesting = false
+            for output,amount in pairs(proposal[2]) do
+               if port.flag.ware_economy:needs(output) then
+                  trade_interesting = true
+                  break
+               end
+            end
+            if trade_interesting then
+               for input,amount in pairs(proposal[1]) do
+                  if port:get_wares(input) < amount then
+                     trade_interesting = false
+                     break
+                  end
+               end
+               if trade_interesting then
+                  for input,amount in pairs(proposal[1]) do
+                     port:set_wares(input, port:get_wares(input) - amount)
+                  end
+                  for output,amount in pairs(proposal[2]) do
+                     port:set_wares(output, port:get_wares(output) + amount)
+                  end
+                  trade_count = trade_count + 1
+                  if trade_count == 3 then
+                     local prior_center = scroll_to_field(map:get_field(90, 132))
+                     sleep(1000)
+                     campaign_message_box(training)
+                     scroll_to_map_pixel(prior_center)
+                     prior_center = nil
+                  end
+                  sleep(3600)
+               end
+            end
+         end
+      end
+      sleep(7200)
+   end
+end
+
+function check_objective_wood()
+   local o = add_campaign_objective(obj_wood)
+   while not check_for_buildings(p1, {frisians_woodcutters_house = 3, frisians_foresters_house = 4}) do
+      sleep(4273)
+   end
+   set_objective_done(o)
+   campaign_message_box(secured_wood)
+   while count_in_warehouses("log") < 80 do
+      sleep(4887)
+   end
+   local prior_center = scroll_to_field(map:get_field(95, 156))
+   sleep(1000)
+   campaign_message_box(diking)
+   scroll_to_map_pixel(prior_center)
+   prior_center = nil
+end
+
+function check_objective_block()
+   local o = add_campaign_objective(obj_block)
+   while true do
+      sleep(4916)
+      local sites = 0
+      for x = 84, 94 do
+         for y = 114, 122 do
+            local f = map:get_field(x, y)
+            if f.immovable ~= nil and f.immovable.descr.type_name == "militarysite" and f.immovable.fields[1] == f and f.owner == p1 then
+               sites = sites + 1
+               if sites >= 3 then
+                  set_objective_done(o)
+                  campaign_message_box(secured_block)
+                  return
+               end
+            end
+         end
+      end
+   end
+end
+
+function check_objective_uplands()
+   local o = add_campaign_objective(obj_uplands)
+   while true do
+      sleep(4471)
+      local nr = 0
+      for _,descr in pairs(p1.tribe.buildings) do
+         for _,b in pairs(p1:get_buildings(descr.name)) do
+            if b.fields[1].height > 40 then
+               nr = nr + 1
+               if nr >= 15 then
```

Comment:
```suggestion
               if nr >= 20 then
```
15 buildings are to few to call it an industry, especially as  military buildings count as well.